packages/core/src/compiler/Style.test.ts

```diff
@@ -56,17 +59,65 @@ export const loadProgs = ({
 }: Trio): {
   env: Env;
   subEnv: SubstanceEnv;
-  subProg: SubProg<A>;
   styProg: StyProg<C>;
+  translation: Translation;
+  state: State;
 } => {
-  const env: Env = envOrError(dsl);
-  const [subEnv, varEnv]: [SubstanceEnv, Env] = subEnvOrError(sub, env);
-  const styProg: StyProg<C> = unsafelyUnwrap(S.parseStyle(sty));
+  const throwErr = (e: any): any => {
+    throw Error(
+      `Expected Style program to work without errors. Got error: ${showError(
+        e
+      )}`
+    );
+  };
+  const env: Env = compileDomain(dsl).unwrapOrElse(throwErr);
+  const [subEnv, varEnv]: [SubstanceEnv, Env] = compileSubstance(
+    sub,
+    env
+  ).unwrapOrElse(throwErr);
+
+  const styProg = S.parseStyle(sty).unwrapOrElse((e) => throwErr(e));
+  const assignment = S.buildAssignment(varEnv, subEnv, styProg);
+  // second pass: construct a dependency graph among those expressions
+  const graph = S.gatherDependencies(assignment);
+
+  const rng = seedrandom("style test");
+  const varyingValues: number[] = [];
+  const inputs: InputMeta[] = [];
+  const makeInput = (meta: InputMeta) => {
+    const val = "pending" in meta ? meta.pending : meta.sampler(rng);
+    const x = input({ key: varyingValues.length, val });
+    varyingValues.push(val);
+    inputs.push(meta);
+    return x;
+  };
+
+  const canvas = S.getCanvasDim("width", graph)
+    .andThen((w) =>
+      S.getCanvasDim("height", graph).map((h) => makeCanvas(w, h))
+    )
+    .unwrapOrElse(throwErr);
+
+  // third pass: compile all expressions in topological sorted order
+  const translation = S.translate(
+    { makeInput },
+    canvas,
+    graph,
+    assignment.diagnostics.warnings
+  );
+
+  const state = S.compileStyle(
```

Comment:
I haven't really copied the entirety of `compileStyle` in here so this is technically running things twice. Wondering if there's a better way to do this.

Reply:
Yeah, I basically just want to run the compiler up to a certain pass. @liangyiliang had some tests that ran it all the way, but many others only need the translation.